### ISSUE TITLE
Zoom In/Out loses keyboard focus

### ISSUE THREAD

**Issue by webstandards** [Leaflet/Leaflet#3159]
Accessibility Finding: On the map, when a keyboard user action the Zoom In or Zoom Out, focus is removed from the map. Recommendation: Don't lose the focus.

http://www.w3.org/TR/WCAG20/#consistent-behavior
3.2.2 On Input: Changing the setting of any user interface component does not automatically cause a change of context unless the user has been advised of the behavior before using the component. (Level A)

So far I've verified against the master branch using the quick start html example. You can tab to the zoom in/out controls, however when you select one of the controls (by hitting enter) it will zoom in/out one level (good), however the focus on the zoom in/out button is totally lost and your keyboard focus is taken back to the start of the map. 

One observation: In the quick start example there is a Marker you can tab to focus and enter to open the Marker popup info, hit enter again and the popup info is closed and focus remains on the Marker. Just starting to dig through the codebase to see what the difference is between the keyboard handling of a Marker vs a Zoom button.

**Comment by mourner** [comment]
You're right, this should be fixed.

**Comment by mlegoff** [comment]
Hello,

I found the way to fix this bug. I just delete this line (69) on ControlZoom.js : .on(link, 'click', this._refocusOnMap, this); 
But it seems to me that it'a little bit simple. Maybe this line is used for something else ?
I make a little demo here : 
http://kara.62.free.fr/zoomfocus/

**Comment by mourner** [comment]
Yeah, originally this was for #2148. I think the right solution is to detect if it was actually clicked by a mouse or a keyboard in the `_refocusOnMap` handler.

**Comment by juliensoret** [comment]
Agree on checking click origin.
User who clicks with mouse doesn't require keeping focus on the button. He's more likely to use arrows as said in https://github.com/Leaflet/Leaflet/issues/2148.
While people who got to zoom by keyboard will surely understand they have to focus back on the map to use arrows.
Or I don't know the difficulty to make the arrows pressing handled by the map while giving back focus to the map.

**Comment by webstandards** [comment]
Had my accessibility consultant review @mlegoff demo site looks good, and he also says ok to make the change only for keyboard users and leave current mouse behaviour alone.

**Comment by Jakobud** [comment]
@mourner I could probably handle this one. Is it possible to pass the event into `_refocusOnMap` and if so, how do you determine if it was a keyboard click or mouse click for the event?

**Comment by mourner** [comment]
@Jacobud that's what you would have to figure out :)

**Comment by Jakobud** [comment]
It looks like the event comes up as a MouseEvent if I click it or if I use the keyboard and press enter.

The only differentiating things between the events is that the actual mouse click event has X,Y values for where the X,Y coordinate was where you clicked. But using the keyboard all those related values as 0,0. In your opinion is it hacky to say if x==0 && y==0 then it's a keyboard, else its a mouse?

Is there are better way to determine if it's a keyboard event?

**Comment by Jakobud** [comment]
I put together something that works. Not sure if its the best solution or not. I don't know if there is a perfect way to determine the difference between a Mouse and Keyboard event or not.

**Comment by mourner** [comment]
Should be fixed, @webstandards @juliensoret please make sure it works as expected.

**Comment by juliensoret** [comment]
It works fine for me !
Here is a demo so anyone can check quickly : http://kara.62.free.fr/zoomfocus_3275/

**Comment by webstandards** [comment]
Did some testing across FF/Chrome/IE11 with NVDA works great. Thanks to everyone who contributed to this, keyboard / screen reader / assistive device users will appreciate it!
